README.md

```diff
@@ -1 +1,166 @@
+            outputs = model(inputs)
+            _, preds = torch.max(outputs, 1)
+            loss = criterion(outputs, labels)
+
+            # zero the parameter gradients
+            optimizer.zero_grad()
+            loss.backward()
+            optimizer.step()
+            count += 1
+
+            # statistics
+            rh.submit_result(
+                f"worker/{worker_rank}/data-{dt}.json",
+                json.dumps({'loss': loss.item(), 'epoch': epoch, 'count': count, 'worker': worker_rank})
+            )
+            if (count % 100) == 0 and worker_rank == 0:
+                rh.submit_result(f"checkpoint-{dt}.pkl", pickle.dumps(model.state_dict()))
+
+```
+
+## How does it work?
+
+`dask-pytorch` is largely a wrapper around existing `pytorch` functionality.  `pytorch.distributed` provides infrastructure for [Distributed Data Parallel](https://pytorch.org/tutorials/intermediate/ddp_tutorial.html).
+
+In DDP, you create N workers, and the 0th worker is the "master", and coordinates the synchronization of buffers and gradients.  In SGD, gradients are normally averaged between all data points in a batch.  By running batches on multiple workers, and averaging the gradients, DDP enables you to run SGD with a much bigger batch size `(N * batch_size)`
+
+In PyTorch, you set some environment variables to configure the "master" host and port, and then you call `init_process_group` before you start training, and `destroy_process_group` when you are done training.  `dask-pytorch` handles this for you by designating 1 dask worker as the master pytorch worker, and calling `init_process_group` and `desetroy_process_group` around the training function that you provide.
+
+### Multi GPU machines
+`dask_cuda_worker` automatically rotates `CUDA_VISIBLE_DEVICES` for each worker it creates (typically one per GPU).  As a result, your PyTorch code should always start with the 0th GPU.
+
+For example, if I have an 8 GPU machine, the 3rd worker will have CUDA_VISIBLE_DEVICES set to `2,3,4,5,6,7,0,1`.  On that worker, if I call `torch.device(0)`, I will get GPU 2.
+
+## What else?
+
+`dask-pytorch` also implements an S3 based ImageFolder.  More distributed friendly datasets are planned.  `dask-pytorch`  Also implements a basic results aggregation framework so that it is easy to collect training metrics across different workers.  Currently, only `DaskResultsHandler` which leverages `dask` pub sub communication protocols is implemented, but an S3 based result handler is planned.
+
+## Some Notes
+
+Dask generally spawns processes.  PyTorch generally forks.  When using a multiprocessing enabled data loader, it is a good idea to pass the `Fork` multiprocessing context to force the use of Forking in the data loader.
+
+Some Dask deployments do not permit spawning processes.  To override this, you can change the [distributed.worker.daemon](https://docs.dask.org/en/latest/configuration-reference.html#distributed.worker.daemon) setting.
```

Comment:
🤯 I just now got that that was what the damned daemon with child processes error was saying.